pypicloud/access/_ldap.py

```diff
@@ -0,0 +1,391 @@
+"""LDAP authentication plugin for pypicloud."""
+
+
+try:
+    import ldap  # pylint: disable=F0401
+except ImportError:
+    LDAP_ENABLED = False
+else:
+    LDAP_ENABLED = True
+
+import logging
+from functools import wraps
+
+from .base import IAccessBackend
+
+
+def reconnect(func):
+    """If the LDAP connection dies underneath us, recreate it."""
+
+    @wraps(func)
+    def _reconnect(*args, **kwargs):
+        """Inner wrap function to reconnect on failure."""
+
+        try:
+            return func(*args, **kwargs)
+        except ldap.LDAPError:
+            LDAP._connect()
+            return func(*args, **kwargs)
+
+    return _reconnect
+
+
+class LDAP(object):
+    """Handles interactions with the remote LDAP server."""
+
+    @staticmethod
+    def configure(settings=None):
+        """Configures self with the settings dictionary."""
+
+        LDAP._id_field = settings["auth.ldap.id_field"]
+        LDAP._url = settings["auth.ldap.url"]
+        LDAP._service_dn = settings["auth.ldap.service_dn"]
+        LDAP._service_password = settings["auth.ldap.service_password"]
+        LDAP._base_dn = settings["auth.ldap.base_dn"]
+        LDAP._all_user_search = settings["auth.ldap.all_user_search"]
+        LDAP._admin_field = settings["auth.ldap.admin_field"]
+        LDAP._admin_dns = [
+            dn for dn in settings["auth.ldap.admin_dns"].splitlines() if dn
+        ]
+
+        LDAP._connect()
+        LDAP.all_users()
+
+    @staticmethod
+    def _connect():
+        """Initializes the python-ldap module and does the initial bind."""
+
+        LDAP._server = ldap.initialize(LDAP._url)
+        LDAP._server.simple_bind_s(LDAP._service_dn, LDAP._service_password)
+
+    @staticmethod
+    @reconnect
+    def all_users():
+        """Returns a list of all user DNs."""
+
+        if not hasattr(LDAP, "_all_users"):
+            results = LDAP._server.search_s(
+                LDAP._base_dn,
+                ldap.SCOPE_SUBTREE,
+                LDAP._all_user_search,
+            )
+            LDAP._all_users = {}
+            for result in results:
+                if LDAP._id_field in result[1]:
+                    LDAP._all_users[result[1][LDAP._id_field][0]] = result[0]
+
+        return list(LDAP._all_users.values())
+
+    @staticmethod
+    def all_usernames():
+        """Returns a list of all user names."""
+
+        return list(LDAP._all_users.keys())
```

Comment:
It seems like you're relying on `all_usernames()` only ever being called after `all_users()`. Is that guaranteed to be true? It's not obvious to me just from looking at how it's called.